In error messages, print the source fragment responsible for the error

Armael opened the pull request:
This PR generalizes to the standard compiler output the "source highlighting" mechanism that is already present in case of error in a "dumb" toplevel and in the testsuite expect tests.

Examples:
![single_line](https://user-images.githubusercontent.com/315125/46618416-b5d72280-cb1f-11e8-94a1-25bff1864256.png)
![multi_line2](https://user-images.githubusercontent.com/315125/46618433-c5566b80-cb1f-11e8-86ed-813872824bb1.png)

The motivation is to improve error readability: it should now be clearer what the mistake is when looking at the compiler output. AFAIK [elm](https://elm-lang.org/blog/compiler-errors-for-humans), [rust](https://blog.rust-lang.org/2016/08/10/Shape-of-errors-to-come.html) or [bucklescript](https://reasonml.github.io/blog/2017/08/25/way-nicer-error-messages) implement similar error rendering, and were inspiration for this work.

-------

The PR reuses the existing code for quoting source code, making a few improvements:
- Line numbers in a left column
- Color when underlining a single-line location, reusing the color for error/warning
![warning](https://user-images.githubusercontent.com/315125/46618968-1dda3880-cb21-11e8-8f2d-2301db15709c.png)
- Ellipsis in case of multi-line locations with too many lines (10 is the limit)
![multi_line_ellipsis](https://user-images.githubusercontent.com/315125/46619010-3e09f780-cb21-11e8-83a1-53f792fc845c.png)

The code for the highlighting function got a bit more complicated, because we now have to be able to read the source to quote from the input file. The current function reads directly from the lexbuf and only works if it hasn't been refilled in the meantime -- which is reasonable for the toplevel but not so much for a .ml file. There are also a lot of changes in the testsuite, that come in a separate commit.

On the plus side, one consequence of these changes is the merge of three different codepaths for error printing into one, which makes me very happy. Specifically, "batch mode", "dumb toplevel" and "expect test" now use exactly the same code for printing errors.

Commit:
Add source highlighting for errors & warnings in batch mode

Commit:
Update tests

Commit:
Changes

Armael commented on parsing/location.ml:
```diff
@@ -292,90 +292,234 @@ let highlight_terminfo lb ppf locs =
+  (string * int) list
+  =
+  (* Converts a global position to one that is relative to the lexing buffer *)
+  let rel n = n - lb.lex_abs_pos in
```
Hum, I suspect this is in general incorrect, because I'm not taking `lb.lex_start_pos` into account?

gasche replied:
(cc @diml or @let-def)

diml replied:
If I read the code of lexing.ml correctly, I believe it should indeed be something like: `n - lb.lex_abs_pos + lb.lex_start_pos`. Does the current code work?

BTW, I find looking directly into the lexing buffer internals a bit tedious. Personally, I would just keep the whole input as a single string and extract the source code from that string, that would make the code simpler.

Armael replied:
> If I read the code of lexing.ml correctly, I believe it should indeed be something like: n - lb.lex_abs_pos + lb.lex_start_pos.

Yeah, that's what I suspected after reading `lexing.ml` a bit more closely

> Does the current code work?

..yes :) [on the testsuite]

> BTW, I find looking directly into the lexing buffer internals a bit tedious.

Agreed.

> I would just keep the whole input as a single string and extract the source code from that string

Hum, which part of the code would take care of that? (note that the lexbuf business is only really needed for the toplevel, otherwise we can just go read the source file)

diml replied:
The code that processes an input file. But yh, I agree that for plain files it's not an issue. For the toplevel, the code would need to be adapted. Without changing everything, we could keep all the input read since the start of the current loop iteration in a `Buffer.t`.

Armael replied:
@diml I just tried your suggestion, and it breaks the function (either stops quoting the source, or quotes wrong parts of it) :D. So I guess the current implementation is correct?

diml replied:
I guess so. Thanks for trying, at least we know the current one is more likely to be right :)

pmetzger commented:
This is superb! I've wanted this the whole time I've used OCaml!

At the risk of asking to gild the lily, might I suggest also moving towards a more standard error location format as well? The FSF coding standards, which are followed by gcc and llvm and many other compilers, specify that line numbers and columns be printed as

```
sourcefile:lineno:column: message
```
or alternatively as:
```
sourcefile:lineno.column: message
```

See: https://www.gnu.org/prep/standards/html_node/Errors.html

Doing it that way makes it easier for tools to parse the mentioned locations.

That said, if this notion is controversial, please ignore it. I don't want to slow down getting error context printed for the user.

Armael commented:
I think this is orthogonal to this PR (it is not touching the `File ..., Line ..., Characters ...` line) so I would keep it for somewhere else :-).

pmetzger commented:
@Armael Fair enough. But it should be done at some point, and given that so many tests suite error messages are going to have to get changed anyway to deal with the changed output, it would be good to do it all at about the same time.

gasche commented on parsing/location.ml:
```diff
@@ -292,90 +292,234 @@ let highlight_terminfo lb ppf locs =
+let highlight_quote ppf
+    ~(get_lines:
+        start_pos:position -> end_pos:position ->
+      (string (* line *) * int (* beginning of line *)) list)
```
Why not use a record type instead of comments in the type?

Armael replied:
Ehh, why not, but that sounds a bit heavywheight.

gasche replied:
If you feel the need to indicate what the values are for, I think that record or object fields are in order.

Armael replied:
Done.

gasche commented on parsing/location.ml:
```diff
@@ -292,90 +292,234 @@ let highlight_terminfo lb ppf locs =
+    let lines =
+      get_lines ~start_pos:leftmost ~end_pos:rightmost
+      |> List.map (fun (line, s) ->
+        line, try_get_line_number s (s + String.length line), s)
```
We don't expect this part of the compiler to be performance-critical, but I'm still a bit uneasy with the quadratic runtime of this code. Isn't there a linear or nlogn version? (It feels like you are only keeping a set around.)

gasche commented on parsing/location.ml:
```diff
@@ -292,90 +292,234 @@ let highlight_terminfo lb ppf locs =
+        | Some line_nb -> max acc (String.length (string_of_int line_nb))
+        | None -> acc
+      ) 0 lines
+    in
```
If you could separate the display code to go in utils/Misc, that could maybe be reused more easily later? I think what you are really implementing is two-column printing with a separator, with a parametrized justification function. I guess that Arthur would have reused that.

gasche replied:
(Besides this whole `highlight_quote` function breaks the "don't mix different abstraction levels" rule for good code. We shouldn't get from list manipulations of location sets to low-level padding code without a transition, we should have auxiliary functions for the logic of each domain.)

gasche commented on parsing/location.ml:
```diff
@@ -292,90 +292,234 @@ let highlight_terminfo lb ppf locs =
+    | [(line, line_nb_opt, line_start_cnum)] ->
+        (* Single-line error *)
+        Format.fprintf ppf "%a%s@," (prefix ~sep:true) line_nb_opt line;
+        Format.fprintf ppf "%a" (prefix ~sep:false) None;
```
What's the benefit of the `~sep:true ... ~sep:false` protocol over just having a command in-between to print the separator?

Armael replied:
Yeah, I got rid of that.

gasche commented on parsing/location.ml:
```diff
@@ -292,90 +292,234 @@ let highlight_terminfo lb ppf locs =
+            highlighted := false;
+            Format.fprintf ppf "@}"
+          end;
+          if is_start pos then begin
```
I'm again uneasy with the linear lookup here, giving quadratic behavior. We could have a set of start positions.

gasche commented on parsing/location.ml:
```diff
@@ -292,90 +292,234 @@ let highlight_terminfo lb ppf locs =
+        let ellipsed_first, ellipsed_last =
+          let lines_nb = List.length lines in
+          if lines_nb > max_lines then
+            let printed_lines = max_lines - 1 in (* ellipsis uses one line *)
```
the ellipsis

Armael replied:
Fixed, thanks.

gasche commented on parsing/location.ml:
```diff
@@ -292,90 +292,234 @@ let highlight_terminfo lb ppf locs =
+  match !input_lexbuf with
+  | Some lb ->
+      begin match lexbuf_lines_around lb ~start_pos ~end_pos with
+      | [] -> (* The input is likely not in the lexbuf anymore *)
```
eh, why not return a `line list option`?

Armael replied:
Then `Some []` would also be an error case, to be handled exactly the same way as `[]`. I don't think it's useful to complicate the logic and distinguish between two failure cases that are going to be handled the same way.

gasche commented on parsing/location.ml:
```diff
@@ -417,25 +561,55 @@ type report_printer = {
+let is_quotable_loc loc =
+  not (is_dummy_loc loc)
+  && loc.loc_start.pos_fname = !input_name
+  && loc.loc_end.pos_fname = !input_name
```
What are the sources of non-quotable locations? Lexer line directives? Inlining in the compiler? Errors spanning both a `.ml` and a `.mli` file?

Armael replied:
I think the two only cases are either:
- dummy locs / ghost locs leaking out of the compiler or a buggy ppx
- if you feed a `.ml` file to the compiler which contains lexer line directives indicating an other file. I think this is relatively rare -- in particular this is not what happens when using -pp or -ppx or a ppx driver. In that case, we do not try to quote the source code, since we do not know where that file might be, so I think it's better to not try and read random files on the filesystem.

gasche replied:
It could be nice to indicate these two scenarios in a comment, to store our understanding of what can happen for posterity.

Armael replied:
Done.

gasche reviewed (commented):
This looks very nice, thanks! Some comments from a first read.

Armael commented:
Thanks for the quick first review @gasche . I agree that the code grew into an untasteful mess in some places, I'll work on cleaning that up today.

Armael commented:
@gasche I just pushed a commit that moves the code for printing two columns into `Misc`, and also tries to address the "quadratic behavior" issues.

I'm not completely satisfied with the second part, though. The new code works by defining a structure for a "bag of start/end positions", created from a list of locations, and by providing cursors that allow traversing this bag in O(l) where l is the number of locs that have been put in the bag.
This sounded nice in principle, but the code was a bit tricky to get right (cursors are tricky it seems!), and feels a bit overkill. Especially since what you call quadratic is in fact O(n*l) where n is the number of characters in the output, and l the number of locations -- and l is in practice 1 or 2 at the moment.

But even ignoring the complexity, I agree the previous code was ugly and had to be rewritten...
I'll let you judge and comment on the new one.

gasche commented on parsing/location.ml:
```diff
@@ -292,90 +413,185 @@ let highlight_terminfo lb ppf locs =
+          List.iter (function
+            | `Start, _ -> Format.fprintf ppf "@{<%s>" highlight_tag
+            | `End, _ -> Format.fprintf ppf "@}"
+          ) ps;
```
Does that really work correctly when there are multiple overlapping locations positions? I might, but it looks a bit subtle.

gasche commented:
I agree that the cursor approach is a bit weird. From the change to the Bag structure, I like the impact on the Location code (nice factorization), not so much the cursor API, and the implementation indeed looks complex. Also, I don't want to waste your time over-designing an efficient solution when there is no strong need for now.

So I would suggest a compromise:
- pick a simpler, nicer API that *can* be efficiently implemented (and whose use gives nicely readable code)
- do a naive implementation based on your knowledge that the set of locations will always remain very small

My suggestion would be : imagine that you have an efficient implementation of an interval tree. The natural API would be to ask for leftmost and rightmost interval, and also whether at least one interval in the bag overlaps with a given absolute position (and maybe to request the list of overlapping intervals). Then implementing this with lists as you suggest could be ok.

(Another approach would be to consider that we don't need to represent overlapping intervals in the structure, as they can always be fused at structure-construction time. Then a Map indexed on the starting position would be enough. But I don't remember if that is enough for your needs, and I'm not sure that trying to be both clever and specialized at the same time is a good idea.)

gasche commented on driver/pparse.ml:
```diff
@@ -184,6 +184,7 @@ let file_aux ~tool_name inputfile (type a) parse_fun invariant_fun
         seek_in ic 0;
         let lexbuf = Lexing.from_channel ic in
         Location.init lexbuf inputfile;
+        Location.input_lexbuf := Some lexbuf;
```
I asked myself about the impact of this change on external users, but I thought about it more and I don't see anything wrong with it, in fact it was probably a bug not to have it there -- in the work we did on expect-style tests we have a similar setting that was necessary for us.

gasche commented:
In principle I approve of the change; we need to converge on a version the code in location.ml that @Armael and myself feel good about.

Commit:
update .depend

Armael commented:
Take 3. (the first commit is from #2103, I'll rebase one it is merged)
I think the set-of-locations business is reasonable now.

There is one issue remaining: check-typo (rightfully) complains about the tests because some of them now use more than 80 columns. Indeed, if an error happens in a 80 col line of code, the error message will contain `42 | theline` which is more than 80 characters long.
What should we do?

gasche commented on parsing/location.ml:
```diff
@@ -225,6 +225,89 @@ let print_locs ppf locs =
+  val of_intervals : (('a * int) * ('a * int)) list -> 'a t
+
+  val mem : 'a t -> pos:int -> bool
+  val find_in : 'a t -> s:int -> e:int -> 'a option
```
At first I couldn't tell what `s` and `e` were for here. Maybe just `int * int`, or more explicit label names.

The specification, if I understand the implementation correctly, is: return the value associated to one of the interval bounds that lies within the `(s, e)` parameters. Maybe `find_bound_in`?

trefis replied:
Agreed, `s` and `e` are just bad names.
More explicit label names are indeed required (and I'd be in favor adding a label even if the function is going to take a pair of `int`s).

Armael replied:
Should be better now (`val find_bound_in : 'a t -> range:(int * int) -> 'a bound option`).

gasche commented on Changes:
```diff
@@ -121,6 +121,9 @@ Working version
 ### Compiler user-interface and warnings:
 
+- PR#????: Add source highlighting for errors & warnings in batch mode
+  (Armaël Guéneau, review by ?)
```
That's Jérémie and me.

gasche commented on parsing/location.ml:
```diff
@@ -225,6 +225,89 @@ let print_locs ppf locs =
+end
+=
+struct
+  type 'a boundary = 'a * int
```
In fact you could expose this type and have `find_boundary_in` return it.

gasche commented on parsing/location.ml:
```diff
@@ -225,6 +225,89 @@ let print_locs ppf locs =
+   overlap).
+*)
+
+module IMapSet : sig
```
Why IMapSet?

gasche commented on parsing/location.ml:
```diff
@@ -225,6 +225,89 @@ let print_locs ppf locs =
+        | `E, `Inside (s, n) -> `Inside (s, n-1), acc
+      ) (`Outside, []) pos in
+    if nesting <> `Outside then
+      raise (Invalid_argument "ISet.of_intervals: invalid intervals");
```
When are intervals `invalid` in this sense?

Should the exception name be changed?

gasche replied:
If I read the implementation correctly, I don't see how this exception can ever be raised given that the bounds only inserted when the `S is before the `E. (So we have never read more Es than Ss, and we have always read exactly at much at the end). So it could be an `assert (nesting = `Outside)`.

Armael replied:
Indeed, the exceptions were actually impossible cases. I turned them into `assert false`.

gasche commented on parsing/location.ml:
```diff
@@ -225,6 +225,89 @@ let print_locs ppf locs =
+   implementation if needed.
+
+   Note: the structure only stores maximal intervals (that therefore do not
+   overlap).
```
Out of curiosity, does the "maximal intervals only" aspect matter much? (It adds some complexity in `of_intervals`, and if I understand correctly it means that some of the user-provided bound annotations are lost.)

Armael replied:
Right. I think it simplifies the logic for inserting `@{<..>` and `@}` later, which will get trickier if it has to handled possibly overlapping intervals (and I think the previous implementation which tried to do it was incorrect).

So I think it's better to merge intervals directly in the interval set structure, additionally that's what an actual interval set implementation would do I believe.

gasche commented on parsing/location.ml:
```diff
@@ -292,90 +375,184 @@ let highlight_terminfo lb ppf locs =
+        Format.fprintf ppf "%s | %s@," line_nb line;
+        Format.fprintf ppf "%*s   " (String.length line_nb) "";
+        for pos = line_start_cnum to rightmost.pos_cnum - 1 do
+          (* NB: The two cases below are mutually exclusive *)
```
There are four cases below, which two do we mean? Maybe use `else if` if the tests are exclusive?

Armael replied:
Indeed, that wasn't clear. I used an `else if` as you suggest.

gasche commented on parsing/location.ml:
```diff
@@ -292,90 +375,184 @@ let highlight_terminfo lb ppf locs =
+   It first tries to read from [!input_lexbuf], then if that fails (because the
+   lexbuf no longers contains the input we want), it reads from [!input_name]
+   directly *)
+let get_lines_from_current_input ~start_pos ~end_pos =
```
`lines_around_from_lexbuf`, `lines_around_from_file`, `lines_around_from_current_input`?

Armael replied:
Done.

gasche reviewed (commented):
I still have some minor comments, but overall this looks good, thanks! The other PR has been merged so you can rebase now.

For line limits, check-typo distinguishes "long lines" (>80) and "very long lines" alarms (>130), so we could conceivably say that long lines that are not very long are ok, or add an "overflowing_line" category at 85. None of this is very hard, but correctly adding this alarm-marker to all testsuite files is tricky for now, before #1910 is merged (and I'm the one that has to turn the crank there).

Another approach would be to print lines that are two long differently:

```
 1 | foo bar
 2 |
very long foo bar
 3 | foo baz
```

gasche commented:
I'm getting fairly satisfied with the implementation, but here is an UI question: suppose a user strongly dislikes the new error formatting (maybe the extra lines break their workflow in some way), is there an easy way for them to disable it?

gasche commented:
(If there is, maybe it could be mentioned in the Changes entry.)

Armael commented:
I think I addressed all of your comments on the code!

I still have to look at the long lines issue. I'm not so fond of the idea of printing long lines differently, it does not look too good, and it only makes a difference for lines between 76 (or so) and 80 characters long...

It does not look too hard to set the long-line attribute on all testsuite files, maybe I could do that?
Or maybe, reformat the code of the tests so that the error message do not overflows anymore?

Armael commented:
> suppose a user strongly dislikes the new error formatting (maybe the extra lines break their workflow in some way), is there an easy way for them to disable it?

There isn't, currently. It's not hard to implement, but we have to think carefully if 1) we need it (adding alternative codepaths has a maintenance cost) 2) how to pass this option (as an environment variable? a commandline flag?)

gasche commented:
> It does not look too hard to set the long-line attribute on all testsuite files, maybe I could do that?

It seems reasonable to me, but I think it would be better to wait for #1910 to be merged (hopefully today or tomorrow) to do it.

> There isn't [a way to disable the feature].

I think that an environment variable could be nice, but on the other hand the complexity of adding the configuration-plumbing makes me as uneasy as you. Maybe we could merge as-is and see what is the feedback in trunk testing and beta versions.

lpw25 commented:
I imagine that people who compile from inside an editor or IDE probably don't want this output, since the editor is going to highlight the code in your buffer anyway. So an option to disable it would probably be appreciated.

Armael commented:
Right, although note that anyone using merlin (directly or through LSP) should not be affected (merlin will disable this part of the message directly through the compiler-libs api).

gasche commented:
My feeling is that Leo is speaking the voice of reason on this one.

You could start with just an OCAML_QUOTE_SOURCE environment variable, or if you want the full command-line-parameter dance just look at how `-color` and OCAML_COLOR are handled in the compiler.

pmetzger commented:
FYI, I don't think an option to disable such messages exists in other compilers I use, and almost all of them produce output like this now.

dbuenzli commented:
While my first gut feeling was @lpw25's one I'm not entirely sure --- but would need to live with it to definitively answer the question. Here are a few points to ponder:

1. Your editor can only highlight a *single* error/warning and you may not want to jump through all the locations to assess the nature of a bunch of errors.
2.  The direct vicinity between the error message and the source extract might actually help to figure out the problem more quickly before jumping to the actual location; your eyes don't have to repeatedly travel between two different areas (error panel/source panel).

trefis commented:
> Right, although note that anyone using merlin (directly or through LSP) should not be affected (merlin will disable this part of the message directly through the compiler-libs api). 

Not sure what you mean by that.

One thing I can say is that the easiest for merlin would be to not have to touch anything except setting something like `Clflags.pretty_errors` to `false`.

> You could start with just an OCAML_QUOTE_SOURCE environment variable

Please stop with the goddamn environment variable thread.
If you want to have them alongside some command line flag, fine, but having them as the only option to control the behavior of the compiler seems like a terrible user interface.

trefis commented:
Also, to react on @dbuenzli's message: I am quite certain we don't want it in vim. I'm sure emacs is nice and opens a buffer to display the error properly or whatever. vim will just join everything on a single line that it will display at the bottom of your screen. So displaying the code would actually hide away the error message.

To be clearer: I'm really enthusiastic about that PR (and all the other work @Armael has been doing in the area), but I'm also convinced we want an easy way (cmdline flag) to disable this nice printing. (I weakly think that the nice printing should be on by default though)

gasche commented:
Yes, we want the nice printing to be on by default, but also to provide an easy way for users to opt out.

Armael commented:
> Not sure what you mean by that.

What I had in mind was: define a `report_printer` that disables the quoting part by reusing `Location.batch_mode_printer`, and then set `Location.report_printer` to it.
It is a bit more code than setting a flag, but in that case avoided having the precise logic "disable that part of the message" hardwired in the compiler.

> [environment variable vs cmdline flag]

AFAIU, environment variables have the advantage that you can use them to communicate to the compiler even if you do not control the build system. I suspect there are cases where the IDE wants to read the output of `make` but isn't integrated enough with the build system to pass custom flags?

Armael commented:
In the long run, I'm wondering if the compiler shouldn't emit errors in a machine-readable format, depending on a flag that IDE/editors could pass.

Then, instead of having to add a flag for each one of the parts of the error message that the editor might want to disable, the editor would just not print the bits of the message it does not want.

trefis commented:
> AFAIU, environment variables have the advantage that you can use them to communicate to the compiler even if you do not control the build system

I'm not saying you shouldn't have the environment variable *in addition* of the flag; I can see how they can be useful. But having *just* the env var (which is what @gasche was suggesting) sounds terrible. It makes writing build systems harder: they have to check if the variable is set or not, because if they just set it unconditionally, good luck working around your build system.

gasche commented:
Well if there was just an environment variable, the intent of it would be that only humans set it manually in their build commands to opt-out of this feature in unplanned scenarios, not build systems, so build systems wouldn't have to do anything with it.

Ultimately I don't strongly care what mechanisms are available to disable the feature, but it seems clear that we need at least one. If someone around here is volunteering to implement this in @Armael's stead, please speak up! Otherwise I would suggest to let him do whatever seems reasonable to him to enable that feature, and then send subsequent PRs if you would like to refine the mechanism.

mshinwell commented:
@gasche Has any surveying of users been done to determine whether the majority view is in fact that error messages should be augmented with source fragments by default?  It seems to me like this is a fairly significant change to the user interface of the compiler.  Might there be accessibility issues to consider?

I'm slightly surprised that if we want to make such a major change that we're not adopting @pmetzger 's suggestion of standardising the location format at the same time.  This may seem to be orthogonal, but I would have thought lots of tests might need updating anyway, so could be done at the same time yielding an overall time saving.

Environment variables don't really seem like a good way of controlling this in isolation.  However I'm unsure whether just having a flag is a good idea either.  It seems like it could become an irritation if you wish to have the opposite of the compiler's default as your personal default.  (As an example, when I'm coding in spacemacs on my laptop, I only have a handful of lines devoted to compilation errors at the bottom because there's insufficient screen space.  I think that means I would always need to have this disabled.)

FWIW, my experience of this with llvm is that it hinders more than it helps, although as I just discussed with @trefis this may be because the cases I'm thinking of often involve more than one error message being printed at once.  At that point it just becomes confusing to see what's going on.  I know we're not in a position in OCaml yet to emit multiple warnings/errors at once, but we may be in the future.

gasche commented:
The fact that this feature is used (and indeed not optional as far as I can see) in compilers with even more users than we have makes me somewhat confident that hell is not going to break loose if we enable the feature. At worst, people will complain, and we will refine the mechanism.

Asking for user surveys on interface changes would be a good way to kill any motivation to improve the user interface.

trefis commented:
> Otherwise I would suggest to let him do whatever seems reasonable to him to enable that feature, and then send subsequent PRs if you would like to refine the mechanism.

That seems at odds with your behavior on some other past PRs (e.g. I remember a discussion on some of sliquisters PRs). In particular considering that everyone agrees that there needs to be a way to turn that printing on or off, it seems weird to not require a complete solution to be implemented.

That being said, I'm sure whatever seems reasonable to @Armael will actually be reasonable (i.e. I'm sure he won't *just* implement support for a env variable).

W.r.t. https://github.com/ocaml/ocaml/pull/2096#issuecomment-431042428 : I agree.

> I'm slightly surprised that if we want to make such a major change that we're not adopting @pmetzger 's suggestion of standardising the location format at the same time. This may seem to be orthogonal, but I would have thought lots of tests might need updating anyway, so could be done at the same time yielding an overall time saving.

We've discussed this offline but let's make it public:
- it is orthogonal
- updating tests is no effort at all
- there is a very genuine risk of breaking people's workflow if we change the format in which locations are printed (and @Armael was indeed told off when he did such a change in a previous PR), whereas there doesn't seem to be any such risk with the present PR.

mshinwell commented:
@gasche I think you're maybe exaggerating; I'm not talking about "hell breaking loose".  I'm simply concerned with subtle changes to people's workflows that may cause frustration over time.  Especially for user interface design, I think it is important to consult with users to determine what works for them, and I don't think it's unreasonable for us to ask for some evidence of that beyond "everyone else does it this way".  Perhaps @Armael has already done this, in any case.

I agree with @trefis that a complete solution to the default values and controllability of this feature should be implemented before it is merged.  This change has somewhat more pervasive impact than many of the pull requests we consider.

gasche commented:
> That seems at odds with your behavior on some other past PRs

I'm trying to improve :-)

damiendoligez commented:
Why print only the first and last line numbers? That looks strange.

Armael commented:
I only print the line numbers for lines in which I have a location. This is because line numbers are not consecutive in general.. :)

```ocaml
let x = 1. +. (1 +
# 999 "toto.ml"
     2)

  1 | ..............(1 +
    | # 999 "toto.ml"
999 |      2)
Error: This expression has type int but an expression was expected of type
         float
```

gasche commented:
Note: #1910 is merged, so it should now be very easy to allow long-line for all `/testsuite/tests/**`.

Commit:
Infer missing line numbers

Commit:
Update the tests

Commit:
Add option -error-style and environment variable OCAML_ERROR_STYLE

Commit:
Allow long lines in tests

Armael commented:
I just added some more commits:
- To make @damiendoligez happy and because I was not fully satisfied with my explanation earlier, I implemented a simple heuristic to infer line numbers in most cases. The heuristic is as follows: if we can infer line numbers in sequence that are consistent with the line numbers we know from the locations, then we do so. We should now get line numbers for all the lines as long as no lexer line directives are involved.
- I added the options `-{no-}fancy-errors` and the environment variable `OCAML_FANCY_ERRORS` that can be used to disable the feature
- I allowed long lines in test files

Armael commented:
The name of the option ("fancy errors") is a bit vague in order to leave the door open to other cosmetic changes, that we would not want to display e.g. in IDEs. Then the option could be reused to disable those as well.

gasche commented:
Rather than `-fancy-errors` / `-no-fancy-errors`, I would prefer to have a parametrized flag `-fancy-errors [true|false]`. I know that `X / -no-X` is more common, but having a parameter gives us the possibility to upgrade to finer-grained settings later (eg. "false/none", "simple", "compact", "true/fancy", "json") without changing the interface. (Maybe that suggests using `-error-style` rather than `-fancy-errors`, but I don't have strong opinions on the name.)

gasche commented on driver/optmain.ml:
```diff
@@ -264,6 +266,7 @@ let main () =
         (use 'ocamlopt -depend -help' for details)"];
     Clflags.parse_arguments anonymous usage;
     Compmisc.read_color_env ();
+    Compmisc.read_fancy_errors_env ();
```
Maybe factorize into a single `Compmisc.read_cflags_from_env ()` invocation?

gasche commented on toplevel/opttopmain.ml:
```diff
@@ -72,6 +72,7 @@ let file_argument name =
     let newargs = Array.sub !argv !Arg.current
                               (Array.length !argv - !Arg.current)
       in
+      Compmisc.read_fancy_errors_env ();
```
Why is `read_color...` not called here?

gasche commented on toplevel/opttopmain.ml:
```diff
@@ -266,6 +269,7 @@ let main () =
     | Arg.Bad msg -> Format.fprintf Format.err_formatter "%s%!" msg; exit 2
     | Arg.Help msg -> Format.fprintf Format.std_formatter "%s%!" msg; exit 0
   end;
+  Compmisc.read_fancy_errors_env ();
```
Same `read_color...` curiosity.

gasche commented on toplevel/topmain.ml:
```diff
@@ -176,6 +179,7 @@ let main () =
     | Arg.Help msg -> Printf.printf "%s" msg; exit 0
   end;
   Compenv.readenv ppf Before_link;
+  Compmisc.read_fancy_errors_env ();
```
`read_color...`?

gasche reviewed (commented):
I think you should also document the new command-line options (manual?, manpages).

Armael commented:
> I think you should also document the new command-line options (manual?, manpages).

Ah yes, will do.

> (Maybe that suggests using eg. -error-style rather than -fancy-errors, but I don't have strong opinions on the name.)

I like the idea. So should the option become `-error-style [classic|fancy]`? 

> Why is read_color... not called here?

`-color` does not seem to be a toplevel option (and therefore the environment variable is not read by the toplevel). It would make sense to add it as a toplevel option, actually. I can do that if you want.

dbuenzli commented:
> I like the idea. So should the option become `-error-style [classic|fancy]`?

I don't see anything fancy in your error format, actually that's the way my very famous® compiler course teacher taught me to output them at the beginning of the century. 

I'd rather have something more descriptive, maybe `contextual` ?

gasche commented:
> `-color` does not seem to be a toplevel option

Well this is wrong, given that errors/warning are printed in color in the toplevel as well, and your new format implies more color usage in some cases.

It's fine to deal with this in a separate PR, but a part of the fix may be to homogenize the read-stuff-from-env code in this PR.

Armael commented:
- I've amended the commit adding the option, to implement `-error-style [classic|contextual]`.
- `-color` is now supported by the toplevel, and I merged the environment reading functions as @gasche suggested
- I documented the option in the manpages/manual

Feel free to bikeshed on the docstring for the option, I'm not very good at writing those.

Armael commented:
Btw, how can I check that I didn't mess up the formatting while editing the manpages?

gasche commented:
`man ./man/foo.m` works, I think. (Looking at your manpage, I wonder if there isn't a description-list syntax that you could have used.)

Armael commented:
(I just pushed a small fix to the manpages, a space was missing after `-error-style`)

gasche commented on driver/compmisc.ml:
```diff
@@ -57,20 +57,42 @@ let initial_env () =
+      env_var = "OCAML_ERROR_STYLE";
+    };
+  ]
+
```
I'm not fond of your implementation.

- It is complex (a record with an existential type)
- It hides the data and exposes only poorly-typed actions, which makes it fairly inflexible

I would think of the following interface.

```
type 'a env_reader
val color_reader : color env_reader
val error_style_reader : error_style env_reader
val set_from_env : 'a option ref -> 'a env_reader -> unit
val read_all_flags_from_env : unit -> unit
```

Thinking about it more, I don't see a reason to have this in `Compmisc`, I would move it to `Clflags`.

(This API I proposed does not easily allow compiler-libs users to add their own flags, but I don't think we need to worry about this.)

Armael replied:
> I don't see a reason to have this in Compmisc, I would move it to Clflags.

It's not possible because it would create a circular dependency between Clflags and Location (which is used in `set_from_env` to display a warning).

Apart from that I implemented what you suggest.

gasche replied:
> it would create a circular dependency between Clflags and Location (which is used in set_from_env to display a warning).

What about `(unit, bad_env_variable_payload) result`?

gasche replied:
Thinking about it more:

1. The `result` type solves the dependency problem but it makes the caller-side code cumbersome. Not so nice.
2. What I really like about the idea of moving to `Clflags` was to have the `env_reader` definitions close to their parse functions, etc. It's not nice that the notion of what are the correct error-style settings have to be spilled across two modules. So maybe an acceptable compromise would be to have `set_var_from_env` in Compenv, and the rest in Clflags. That may require either exposing the record type from Clflags, or an internal `try_set_from_env` function with a more complex calling protocol because it is error-handling-parametric (for example the one suggested above).

gasche replied:
Armael wrote

> Right. I exposed env_reader, and moved the reader definitions to Clflags. I also reused the usage message in compenv.ml as you pointed out, which is nice.

This is now resolved.

gasche reviewed (commented):
I made an implementation comment on the flags API.

There is something a bit strange about the current command-line interface: `-error-style <style>` expects either `classic` or `contextual`, and you say that the default value is `contextual`. From this one could infer that not passing the `-error-style <style>` flag is equivalent to passing the `-error-style contextual` flag. But then you say: if you haven't passed the flag, `OCAML_ERROR_STYLE` is consulted.

It may make sense to have a third `auto` value supported as a command-line option (but either unsupported or ignored by the environment-value parser) whose semantics is just to cascade/inherit the environment value (or other future default-guesses we could make), and use `contextual` if absent.

Armael commented:
> It may make sense to have a third auto value supported as a command-line option (but either unsupported or ignored by the environment-value parser) whose semantics is just to cascade/inherit the environment value (or other future default-guesses we could make), and use contextual if absent.

That sounds overly complex. I'm not sure I see the point of introducing a third value for the option which is strictly equivalent to *not* passing the option, and is not supported by the environment variable.
In comparison, with `-color`/`OCAML_COLOR`, all values that are valid for the option are also valid in the environment variable.

gasche commented:
Minor reasons for having this feature exposed (`-error-style auto`) include the abilities to (1) provide a value to say "just listen to the environment" in settings where one has programmatically decided to emit the option anyway and is asking which value to provide the user and (2) override previous settings using by passing another command-line option later in the command.

If you think this is too complex, feel free to follow your intuition and not to do anything. But then I think you should not say that "contextual" is the default in help/usage, or maybe say something like "default, unless OCAML_ERROR_STYLE is used, see manual". To me "X is the default" means "passing nothing is the same as passing X".

Armael commented:
>  (-error-style auto) include the abilities to (1) provide a value to say "just listen to the environment"

Note that this would not be consistent with `-color auto`: passing `-color auto` on the commandline *overrides* the value provided by the environment variable. So it's worth noting `-color` currently does not provide the mechanism you describe, and naming it `-error-style auto` would be extra confusing I think.

I think I'm simply going to mention OCAML_ERROR_STYLE in the usage message (should I do it for `-color` as well, since your remarks also apply in that case?)

Commit:
unified-options.etex: make check-typo happy

Octachron commented on manual/manual/cmds/unified-options.etex:
```diff
@@ -138,6 +140,21 @@ The environment variable "OCAML_COLOR" is considered if "-color" is not
 
+\notop{%
+\item["-error-style" \var{mode}]
+Cnotrol the way error messages and warnings are printed.
```
Typo: Control

Armael replied:
Fixed, thanks.

gasche commented on driver/compenv.ml:
```diff
@@ -350,6 +350,15 @@ let read_one_param ppf position name v =
+      | None ->
+          Printf.ksprintf (print_error ppf)
+            "bad value %s for \"error-style\", \
+             (expected \"contextual\" or \"classic\")" v
```
(Here one could reuse `Clflags.error_style_reader.usage` if the reader type was exposed. Not sure that is important.)

Armael commented:
Right. I exposed `env_reader`, and moved the reader definitions to Clflags. I also reused the usage message in compenv.ml as you pointed out, which is nice.

Armael commented:
And I just amended the docstring of the option to mention the existence of the environment variable.

dbuenzli commented on driver/main_args.ml:
```diff
@@ -597,7 +597,22 @@ let mk_color f =
+  Printf.sprintf
+    "  Control the way error messages and warnings are printed\n\
+    \    The following settings are supported:\n\
+    \      classic     only print the error and its location\n\
```
That will only talk to old timers 👴. Here's a brain dump: `short`, `brief`, `minimal`, `location` ?

gasche replied:
(I'm not fond of "location" but the other suggestions seem fine.)

Armael replied:
Good point. I went with `short`.

gasche commented on parsing/location.ml:
```diff
@@ -657,14 +657,21 @@ let is_quotable_loc loc =
     in
     let highlight ppf loc =
-      if is_quotable_loc loc then
+      if error_style () = Misc.Error_style.Contextual
```
This should be a non-fragile pattern-matching instead of an equality test, to force us to revisit if we add another option.

dbuenzli commented:
> It may make sense to have a third `auto` value supported as a command-line option (but either unsupported or ignored by the environment-value parser) whose semantics is just to cascade/inherit the environment value (or other future default-guesses we could make), and use `contextual` if absent.

@gasche Just to make things a bit clearer. The way usually these kind of cli API is implemented/expected (and supported by e.g. `cmdliner`) is as follows: 

Your program has a setting which has a default value and whose value can be changed on the cli via optional(s) argument(s). If the setting has a corresponding environment variable, setting the environment variable only changes *the default* of the setting and that's all (i.e. a cli specification always "wins").

`auto` in the case of colors is something else. `auto` means that the tool is trying to be smart about choosing the final behaviour (in the case of colors by consulting the result of `isatty(3)` on stdout or stderr) but it is a value for the setting as any other value (and can also be set in the environment variable, though that is useless since `auto` is the default, unless I'm mistaken).

gasche commented on toplevel/opttopmain.ml:
```diff
@@ -241,6 +242,16 @@ module Options = Main_args.Make_opttop_options (struct
+    begin match color_reader.parse option with
+          | None -> ()
+          | Some setting -> color := Some setting
+    end
```
I count 12 occurrences of this code pattern in this patch. Could we factorize this into a Misc function? There is a risk of confusion with the properly-error-reporting Compmisc function, so a longer name should be used for the present one to pay for the dubious ignore semantics.  What about `set_or_ignore : (string -> 'a option) -> 'a option env -> unit`? (Being careful to specify that on `None`, it doesn't set the reference, instead of setting it to None as one could naively expect.)

Armael replied:
Done, I think?

gasche replied:
Seems fine to me, thanks!

gasche reviewed (approved):
I made two minor code comments but in general I'm fairly happy with the shape of the code now. I'm marking this Approved to show progress, although of course we should wait for the error-style naming suggestions to converge before merging.

dbuenzli commented:
I would also like to add that in this particular example build systems should always only go for the default. This allows end-users (or `opam`) to force the format they actually desire.

dbuenzli commented on driver/main_args.ml:
```diff
@@ -597,7 +597,22 @@ let mk_color f =
+    "  Control the way error messages and warnings are printed\n\
+    \    The following settings are supported:\n\
+    \      short       only print the error and its location\n\
+    \      contextual  like \"classic\", but also display the source code\n\
```
`like "classic"` -> `like "short"`

Armael replied:
Argh, indeed, thanks.

dbuenzli commented on driver/main_args.ml:
```diff
@@ -597,7 +597,22 @@ let mk_color f =
+;;
+
+let mk_error_style f =
+  "-error-style", Arg.Symbol (["contextual"; "classic"], f),
```
`"classic"` -> `"short"`

gasche commented:
It is my understanding that all issues have been resolved. Unless I hear otherwise, I plan to merge if CI passes.

(Thanks @dbuenzli for the naming help!)

Armael commented:
Many thanks to everyone involved!

shindere commented:
Commit ef76e6e630fc3740ee32ae03451ddd0508482ecc has just been pushed to
trunk to update two test reference files that had been forgotten.
They are used in `-no-flat-float-array` mode.

mshinwell commented:
@Armael @shindere : This GPR appears to break the testsuite when `OCAML_ERROR_STYLE` is not set to the default.  Could you please fix this for 4.08?

cc @damiendoligez

Armael commented:
What do you suggest? 
By design changing `OCAML_ERROR_STYLE` changes the output of all error messages emitted by the compiler; so indeed it will break all the tests that emit error messages. 
Do you have a specific fix in mind?

Armael commented:
@trefis tells me on IRC that we should just make sure ocamltest explicitly sets the `OCAML_ERROR_STYLE` environment variable to the default.
This sounds fairly easy (and a job for @shindere :-)).

shindere commented:
Armaël Guéneau (2019/02/26 08:26 -0800):
> @trefis tells me on IRC that we should just make sure ocamltest explicitly sets the `OCAML_ERROR_STYLE` environment variable to the default.
> This sounds fairly easy (and a job for @shindere :-)).

What a wonderful way for you @Armael to start diving into the wonderful
code of ocamltest!!! ;-)

trefis commented:
I'd like to point out that I didn't suggest that part:
> This sounds fairly easy (and a job for @shindere :-)).

What I will suggest however is making sure that other variables which can affect the testsuite are handled the same way, e.g. `OCAMLRUNPARAM`.

Armael commented:
@mshinwell I'll try to look at that tomorrow and implement @trefis suggestions (have ocamltest explicitly unset the environment variables that can modify the compiler output).

shindere commented:
Thomas Refis (2019/02/26 08:42 -0800):
> I'd like to point out that I didn't suggest that part:
> > This sounds fairly easy (and a job for @shindere :-)).

Too late, I'm already very angry at you @trefis ;-)

> What I will suggest however is making sure that other variables which
> can affect the testsuite are handled the same way, e.g.
> `OCAMLRUNPARAM`.

That one should be handled correctly.

Armael commented:
See #2271 .
